src/cpp/fastdds/core/condition/WaitSetImpl.cpp

```diff
@@ -111,16 +111,28 @@ ReturnCode_t WaitSetImpl::wait(
 
     auto fill_active_conditions = [&]()
             {
-                bool ret_val = false;
-                active_conditions.clear();
-                for (const Condition* c : entries_)
+                bool ret_val;
+                unsigned int old_counter;
+
+                // Loop if predicate may be outdated
+                do
                 {
-                    if (c->get_trigger_value())
+                    ret_val = false;
+                    old_counter = notifications_;
+                    active_conditions.clear();
+
+                    for (const Condition* c : entries_)
                     {
-                        ret_val = true;
-                        active_conditions.push_back(const_cast<Condition*>(c));
+                        if (c->get_trigger_value())
+                        {
+                            ret_val = true;
+                            active_conditions.push_back(const_cast<Condition*>(c));
+                        }
                     }
                 }
+                while (old_counter != notifications_
+                && active_conditions.size() != entries_.size());
+
                 return ret_val;
```

Comment:
If the condition variable is notified by `wake_up` after `while` and before `return`, it seems that the notification is still lost.
I have to admit that the `AnotherTestCondition` used in the regression test can't cover this case.

Reply:
You are absolutely right! 👍 There is a little window where the notification can be lost.
A regression test for that would be difficult to implement because now any user-provided code as the Condition evaluation cannot increase the notification lost window.